File: packages/shared-types/src/course.ts
export type CourseLevel = "beginner" | "intermediate" | "advanced" | string;
export type ProgressStatus = "not_started" | "in_progress" | "completed" | string;

export type CourseResponse = {
  id: string;
  title: string;
  description: string | null;
  level: CourseLevel;
  sort_order: number;
  is_published: boolean;
  prerequisite_course_id: string | null;
};

export type CourseWithProgressResponse = CourseResponse & {
  progress_status: ProgressStatus;
  progress_pct: string;
  is_unlocked: boolean;
};

export type CourseListResponse = {
  courses: CourseWithProgressResponse[];
};

File: apps/backend/app/schemas/course.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel


class CourseResponse(BaseModel):
    id: str
    title: str
    description: str | None
    level: str
    sort_order: int
    is_published: bool
    prerequisite_course_id: str | None = None

    model_config = {"from_attributes": True}


class CourseWithProgressResponse(CourseResponse):
    progress_status: str = "not_started"
    progress_pct: Decimal = Decimal("0.00")
    is_unlocked: bool = False


class CourseListResponse(BaseModel):
    courses: list[CourseWithProgressResponse]
